storage/tick.go

```diff
@@ -0,0 +1,156 @@
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package storage
+
+import (
+	"sync"
+	"time"
+
+	"github.com/m3db/m3db/clock"
+	"github.com/m3db/m3db/context"
+
+	"github.com/uber-go/tally"
+)
+
+const (
+	defaultTickCheckInterval = time.Second
+)
+
+type tickManagerMetrics struct {
+	tickDuration       tally.Timer
+	tickCancelled      tally.Counter
+	tickDeadlineMissed tally.Counter
+	tickDeadlineMet    tally.Counter
+}
+
+func newTickManagerMetrics(scope tally.Scope) tickManagerMetrics {
+	return tickManagerMetrics{
+		tickDuration:       scope.Timer("tick.duration"),
+		tickCancelled:      scope.Counter("tick.cancelled"),
+		tickDeadlineMissed: scope.Counter("tick.deadline.missed"),
+		tickDeadlineMet:    scope.Counter("tick.deadline.met"),
+	}
+}
+
+type tickManager struct {
+	sync.RWMutex
+
+	database database
+	opts     Options
+	nowFn    clock.NowFn
+	sleepFn  sleepFn
+
+	metrics tickManagerMetrics
+	c       context.Cancellable
+	tokenCh chan struct{}
+}
+
+func newTickManager(database database) databaseTickManager {
+	opts := database.Options()
+	scope := opts.InstrumentOptions().MetricsScope()
+
+	return &tickManager{
+		database: database,
+		opts:     opts,
+		nowFn:    opts.ClockOptions().NowFn(),
+		sleepFn:  time.Sleep,
+		metrics:  newTickManagerMetrics(scope),
+		c:        context.NewCancellable(),
+		tokenCh:  make(chan struct{}, 1),
+	}
+}
+
+func (mgr *tickManager) Tick(softDeadline time.Duration, force bool) bool {
+	acquired := false
+	select {
+	case mgr.tokenCh <- struct{}{}:
+		acquired = true
+	default:
+		// If there is an ongoing tick and ticking is not forced, return immediately
+		if !force {
+			return false
+		}
+	}
+
+	// Otherwise if not acquired, it means this is a forced tick so we attempt to
+	// cancel and re-acquire
+	if !acquired {
+		tick := time.NewTicker(defaultTickCheckInterval)
+		for {
+			select {
+			case mgr.tokenCh <- struct{}{}:
+				acquired = true
+				break
+			case <-tick.C:
+				mgr.c.Cancel()
+			}
+			if acquired {
+				break
+			}
+		}
+	}
```

Comment:
Not sure if I understood what you meant, the code you gave does the same thing as mine if `force` is false, and uses a blocking wait if `force` is true. However, when performing a tick during bootstrapping (i.e., when `force` is true), we'd need to cancel the ongoing tick if any instead of waiting for the current tick to finish to avoid incurring a full ticking delay (currently 10 minutes) for bootstrapping, and I don't see how a blocking wait would support cancellation?